File: Algorithm/342. Power of Four.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-08-11 09:40:30
# @Last modified by:   WuLC
# @Last Modified time: 2016-08-11 09:40:38
# @Email: liangchaowu5@gmail.com

class Solution(object):
    def isPowerOfFour(self, num):
        """
        :type num: int
        :rtype: bool
        """
        count = 0
        while count < 32:
            if num & 1 == 1:
                if num >> 1 == 0 and count %2 == 0:
                    return True
                else:
                    return False
            count += 1
            num >>= 1
        return False

File: Algorithm/Java/342. Power of Four.java
/**
* Author: LC
* Date:   2017-06-11 20:37:17
* Last modified by:   LC
* Last Modified time: 2017-06-11 20:42:12
* Email: liangchaowu5@gmail.com
*/


// bit manipulation
// (num&(-num)) == num is checking that num has only one bit that is not 0
// (num&0x55555555) == num is checking the only one bit that is not 0 is in the correct position
// avoid special case when num = 0
public class Solution 
{
    public boolean isPowerOfFour(int num) 
    {
        return  num != 0 && (num&(-num)) == num && (num&0x55555555) == num;
    }
}